fairlearn/linear_model/_constrained_logistic.py

```diff
@@ -0,0 +1,813 @@
+            accept_sparse="csr",
+            dtype=np.float64,
+            accept_large_sparse=solver not in ["liblinear", "sag", "saga"],
+        )
+        y = check_array(y, ensure_2d=False, dtype=None)
+        check_consistent_length(X, y)
+    _, n_features = X.shape
+
+    classes = np.unique(y)
+
+    multi_class = _check_multi_class(multi_class)
+    if pos_class is None:
+        if classes.size > 2:
+            raise ValueError("To fit OvR, use the pos_class argument")
+        # np.unique(y) gives labels in sorted order.
+        pos_class = classes[1]
+
+    # If sample weights exist, convert them to array (support for lists)
+    # and check length
+    # Otherwise set them to 1 for all examples
+    sample_weight = _check_sample_weight(
+        sample_weight, X, dtype=X.dtype, copy=True
+    )
+
+    # If class_weights is a dict (provided by the user), the weights
+    # are assigned to the original labels. If it is "balanced", then
+    # the class_weights are assigned after masking the labels with a OvR.
+    le = LabelEncoder()
+    if isinstance(class_weight, dict):
+        class_weight_ = compute_class_weight(
+            class_weight, classes=classes, y=y
+        )
+        sample_weight *= class_weight_[le.fit_transform(y)]
+
+    # For doing a ovr, we need to mask the labels first.
+    w0 = np.zeros(n_features + int(fit_intercept), dtype=X.dtype)
+    mask_classes = np.array([-1, 1])
+    mask = y == pos_class
+    y_bin = np.ones(y.shape, dtype=X.dtype)
+    y_bin[~mask] = -1.0
+    # for compute_class_weight
+
+    if class_weight == "balanced":
+        class_weight_ = compute_class_weight(
+            class_weight, classes=mask_classes, y=y_bin
+        )
+        sample_weight *= class_weight_[le.fit_transform(y_bin)]
+
+    if coef is not None:
+        # it must work both giving the bias term and not
+        if coef.size not in (n_features, w0.size):
+            raise ValueError(
+                "Initialization coef is of shape %d, expected shape %d or %d"
+                % (coef.size, n_features, w0.size)
+            )
+        w0[: coef.size] = coef
+
+    target = y_bin
+    func = _logistic_loss
+
+    coefs = list()
+    n_iter = np.zeros(len(Cs), dtype=np.int32)
+    for i, C in enumerate(Cs):
+        if solver == "SLSQP":
+            iprint = [-1, 50, 1, 100, 101][
+                np.searchsorted(np.array([0, 1, 2, 3]), verbose)
+            ]
+            opt_res = optimize.minimize(
+                fun=func,
+                x0=w0,
+                method="SLSQP",
+                # jac=True,  # Commented because it is not used by the authors of the paper
+                args=(X, target, 1.0 / C, sample_weight),
+                options={"iprint": iprint, "maxiter": max_iter},
+                constraints=constraints,
+            )
+            n_iter_i = _check_optimize_result(
+                solver="lbfgs",  # Not the actual solver we are using, but it works fine like this
```

Comment:
What does this do then?

Reply:
Giving any other solver than `lbfgs` will return a `NotImplementedError` (See [here](https://github.com/scikit-learn/scikit-learn/blob/16625450b58f555dc3955d223f0c3b64a5686984/sklearn/utils/optimize.py#L217-L266)), since sklearn assumes no other solver from [`scipy.optimize.minimize`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) will be used. However, `SLSQP` and `lbfgs` both originate from `scipy.optimize.minimize`, making it such that the code works fine for the `SLSQP` solver while passing the `lbfgs` parameter. 

So basically, we can still use the sklearn code for checking the optimze result, but we have to pass the parameter that sklearn expects us to pass. If there would be a better way to go about this, I'm open to that as well!

Reply:
After some experiments and encountering some `ConvergenceWarnings`, I decided to implement `_check_optimize_result` myself. Users could also encounter a `ConvergenceWarnings`, and I think an error message with the wrong solver would not be very useful to them in that case.